node_modules/col-activities/lib/notifications/daily.js

```diff
@@ -0,0 +1,548 @@
+  } else {
+    subject = 'Today\'s activity in the Asset Library and Whiteboards';
+    if (!course.canvas.name) {
+      subject = 'Today\'s activity'
+    }
+  }
+
+  return subject;
+};
+
+/* Activity aggregation */
+
+/**
+ * Get the relevant activities for a user
+ *
+ * @param  {Object}       data    The relevant course data
+ * @param  {User}         user    The user for which to get the activities
+ * @return {Activity[]}           A set of activities for which the user can be notified
+ * @api private
+ */
+var getActivitiesForUser = function(data, user) {
+  // This variable will contain all activities that can be emailed to the user
+  var emailActivities = [];
+
+  // Asset comments
+  _.each(data.assets, function(asset) {
+    // Whether the user owns the asset
+    var isOwner = asset.users[user.id];
+
+    // All top comments made by other users
+    var topOtherComments = _.filter(asset.comments, function(comment) {
+      return (!comment.parent && comment.user_id !== user.id);
+    });
+
+    // All replies made to the current user
+    var replies = _.filter(asset.comments, function(comment) {
+      return (comment.parent && comment.parent.user_id === user.id);
+    });
+
+    // All replies made to the current user by other users
+    var otherReplies = _.filter(replies, function(reply) {
+      return (reply.user_id !== user.id);
+    });
+
+    // If we own the asset, we want all the new comments (that weren't made by ourselves)
+    if (isOwner && !_.isEmpty(topOtherComments)) {
+      emailActivities.push(new AssetActivity(asset, 'asset_comment'));
+    }
+
+    // If there's a reply on a comment of ours we only want those replies
+    else if (!_.isEmpty(replies) && !_.isEmpty(otherReplies)) {
+      emailActivities.push(new AssetActivity(asset, 'asset_comment_reply', replies));
+    }
+  });
+
+  // Add whiteboards that have new chat messages and the user is a member of
+  _.each(data.whiteboards, function(whiteboard) {
+    var otherChatMessages = _.filter(whiteboard.chats, function(chat) {
+      return (chat.user_id !== user.id);
+    });
+
+    if (_.find(whiteboard.users, {'id': user.id}) && !_.isEmpty(otherChatMessages)) {
+      emailActivities.push(new WhiteboardActivity(whiteboard))
+    }
+  });
+
+  return _.chain(emailActivities)
+    .sortByOrder(['lastActivity'], ['desc'])
+    .value();
+
+};
+
+/**
+ * An Asset activity
+ *
+ * @param {Asset}       asset       The asset on which there's some activity
+ * @param {String}      type        One of `asset_comment` or `asset_comment_reply`
+ * @param {Comment[]}   [replies]   In case it's an `asset_comment_reply` activity, this should be the set of comments that are replies to a comment of the current user
+ */
+var AssetActivity = function(asset, type, replies) {
+  // Take a copy of the asset
+  asset = _.extend({}, asset);
+
+  if (type === 'asset_comment_reply') {
+    asset.comments = replies;
```

Comment:
No, the passed in `asset` object should not be changed

```javascript
$ node
> var lodash = require('lodash');
undefined
> var asset = {'comments': [1, 2, 3, 4, 5]}
undefined
> var ptr = asset // Create a pointer to the original object
undefined
> asset
{ comments: [ 1, 2, 3, 4, 5 ] }
> ptr
{ comments: [ 1, 2, 3, 4, 5 ] }
> var replies = [11, 12]
undefined
> asset = lodash.extend({}, asset)    // extend will copy over the entries from asset to a new object. This will be returned and assigned to the asset variable.
{ comments: [ 1, 2, 3, 4, 5 ] }
> asset
{ comments: [ 1, 2, 3, 4, 5 ] }
> ptr // Note that ptr points to the ORIGINAL object, whilst asset is now a completely new object
{ comments: [ 1, 2, 3, 4, 5 ] }
> asset.comments = replies
[ 11, 12 ]
> asset
{ comments: [ 11, 12 ] }
> ptr
{ comments: [ 1, 2, 3, 4, 5 ] }
```